src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorSyntaxTreeExtensions.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Formatting
+{
+    internal static class RazorSyntaxTreeExtensions
+    {
+        public static IReadOnlyList<FormattingSpan> GetFormattingSpans(this RazorSyntaxTree syntaxTree)
+        {
+            if (syntaxTree == null)
+            {
+                throw new ArgumentNullException(nameof(syntaxTree));
+            }
+
+            var visitor = new FormattingVisitor(syntaxTree.Source);
+            visitor.Visit(syntaxTree.Root);
+
+            return visitor.FormattingSpans;
+        }
+
+        public static RazorDirectiveSyntax[] GetCodeBlockDirectives(this RazorSyntaxTree syntaxTree)
+        {
+            if (syntaxTree == null)
+            {
+                throw new ArgumentNullException(nameof(syntaxTree));
+            }
+
+            var codeBlockDirectives = syntaxTree.Root
```

Comment:
Might be nice to have a short comment explaining exactly why this logic is the way it is.